File: 98.validate-binary-search-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	if root == nil {
		return true
	}
	return isValidBSTSub(root, -1, -1)
}

func isValidBSTSub(root *TreeNode, lower, upper int) bool {
	if lower != -1 && root.Val <= lower {
		return false
	}
	if upper != -1 && root.Val >= upper {
		return false
	}
	lValid, rValid := true, true
	if root.Left != nil {
		lValid = isValidBSTSub(root.Left, lower, root.Val)
	}
	if root.Right != nil {
		rValid = isValidBSTSub(root.Right, root.Val, upper)
	}
	return lValid && rValid
}

File: 98.validate-binary-search-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        return this.f(root, -Math.pow(2, 31) - 1, Math.pow(2, 31));
    }

    private boolean f(TreeNode root, double min, double max) {
        if (root == null) {
            return true;
        }
        if (root.val <= min || root.val >= max) {
            return false;
        }
        return this.f(root.left, min, (double) root.val) && this.f(root.right, (double) root.val, max);
    }
}